File: ios-native/Jottery/Jottery/Services/HashChainService.swift
import Foundation

/// Content hash computation for sync conflict detection.
enum HashChainService {

    /// Compute the content hash for a decrypted note's encrypted form.
    /// The note must already have been re-encrypted before calling this.
    static func computeContentHash(
        encryptedContent: String,
        encryptedTags: [String],
        attachmentIds: [String],
        pinned: Bool,
        archived: Bool,
        locked: Bool,
        syntaxLanguage: String?,
        wordWrap: Bool,
        showPreview: Bool,
        color: String?
    ) -> String {
        let fields = SyncableNoteFields(
            content: encryptedContent,
            tags: encryptedTags,
            attachmentIds: attachmentIds,
            pinned: pinned,
            archived: archived,
            locked: locked,
            syntaxLanguage: syntaxLanguage,
            wordWrap: wordWrap,
            showPreview: showPreview,
            color: color
        )
        return CryptoService.computeContentHash(for: fields)
    }

    /// Update hash chain by prepending new hash.
    static func updateHashChain(currentHash: String, parentChain: [String]) -> [String] {
        CryptoService.updateHashChain(currentHash: currentHash, parentChain: parentChain)
    }

    /// Find common ancestor between two chains.
    static func findCommonAncestor(chainA: [String], chainB: [String]) -> String? {
        CryptoService.findCommonAncestor(chainA: chainA, chainB: chainB)
    }
}

File: android-native/app/src/main/java/com/jottery/android/service/HashChainService.kt
package com.jottery.android.service

import com.jottery.android.crypto.CryptoService
import com.jottery.android.model.SyncableNoteFields

/**
 * Hash chain management for sync conflict detection.
 * Delegates to CryptoService for the actual hash computations.
 */
object HashChainService {

    /** Compute the content hash for a note's current state. */
    fun computeContentHash(fields: SyncableNoteFields): String {
        return CryptoService.computeContentHash(fields)
    }

    /** Update hash chain by prepending the new hash. */
    fun updateHashChain(currentHash: String, parentChain: List<String>): List<String> {
        return CryptoService.updateHashChain(currentHash, parentChain)
    }

    /** Find a common ancestor between two hash chains. */
    fun findCommonAncestor(chainA: List<String>, chainB: List<String>): String? {
        return CryptoService.findCommonAncestor(chainA, chainB)
    }

    /** Determine if chain B is a descendant of chain A (A's head is in B). */
    fun isDescendant(ancestorChain: List<String>, descendantChain: List<String>): Boolean {
        if (ancestorChain.isEmpty()) return true
        val ancestorHead = ancestorChain.first()
        return ancestorHead in descendantChain
    }
}
